### PR TITLE
Add extension point to notify when browse mode state becomes active or inactive

### LINKED ISSUE #14969: extensionPoints: Notify when NVDA is in browse mode
### Is your feature request related to a problem? Please describe.
Add-ons may need to be notified when NVDA is in browse mode, for example to activate and deactivate profiles when entering and exiting browse mode, respectively.
### Describe the solution you'd like
Add extensionPoints to notify when NVDA is in browse modeor not (i.e., in focus mode or not in a treeInterceptor).
### Describe alternatives you've considered
Adda profile trigger for browse mode, which may take priority over the profile application trigger.
### Additional context
Sometimes maybe useful to change voice settings to read in browse mode. This has been requested by a member of the spanish community on a Telegram group.

### PR DESCRIPTION
**Pull request by nvdaes**
- Add extension point to notify browseMode changes
- Update developer guide

### Link to issue number:
fixes #14969
### Summary of the issue:
Sometimes, it would be desirable perform actions when browse mode state changes. For example, this may be useful to activate or deactivate a profile when entering or exiting browse mode.
### Description of user facing changes
- Add-ons and other components can be notified about browse mode state changes.
### Description of development approach
A treeInterceptorHandler.post_browseModeStateChange extension point has been added. The browseMode parameter represents the state of browse mode (True or False).
### Testing strategy:
- NVDA seems not to produce additional errors.
- Checked that a profile can be activated/deactivated programatically when browseMode is changed with NVDA+space, navigating the web, and  opening the menu bar and other applications.
### Known issues with pull request:
None.
### Change log entries:

For Developers
Added treeInterceptorHandler.post_browseModeStateChange extension points, to be notified about state changes in browse mode (#14969)
### Code Review Checklist:

- [X] Pull Request description:
  - description is up to date
  - change log entries
- [X] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [X] API is compatible with existing add-ons.
- [X] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [X] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [X] Security precautions taken.

### CODE REVIEW

**Comment by nvdaes** [comment]
cc: @leonardder @XLTechie

File: source/api.py [hunk 1]
```diff
@@ -164,6 +164,11 @@ def setFocusObject(obj: NVDAObjects.NVDAObject) -> bool:  # noqa: C901
 		obj.treeInterceptor=treeInterceptorObject
 	else:
 		obj.treeInterceptor=None
+	if obj.treeInterceptor:
```
Comment (leonardder) [inline source/api.py:167]: Isn't this block executed on every focus change? I guess that's problematic.

**Comment by nvdaes** [comment]
> Isn't this block executed on every focus change? I guess that's problematic.

Seems not to have undesirable effects. Anyway, if you have another suggestion, please let me know.

**Comment by leonardder** [comment]
I'd rather see the extension point only called on a change.
May be it should be called in browseMode.reportPassThrough instead? This will however not be notified when moving in and out of a window that has browse mode enabled.

**Comment by nvdaes** [comment]
I'd rather see the extension point only called on a change.
May be it should be called in browseMode.reportPassThrough instead? This will however not be notified > when moving in and out of a window that has browse mode enabled.

For this reason, for now the best way to achieve the possibility of being notified about changes is this one. Hope a best way can be found if this is problematic.

File: source/treeInterceptorHandler.py [hunk 2]
```diff
@@ -20,10 +20,17 @@
...
 
 if TYPE_CHECKING:
 	import NVDAObjects
 
+# Extension points action:
+# Triggered when browseMode state is changed.
+# Callback definition: Callable(browseMode: bool) -> None
+# browseMode - True when current treeInterceptor.passThrough is False.
+# False otherwise, or if focus is not in a treeInterceptor.
+post_browseModeStateChange = extensionPoints.Action()
```
Comment (seanbudd) [inline source/treeInterceptorHandler.py:33]: can this get more detailed documentation on usage in-line with other extension points?

For example see `pre_writeCells`

File: source/api.py (hunk 1 above)
Comment (seanbudd) [reply source/api.py:167]: Agreed, this needs needs to be changed. It's probably worth caching the previous browse mode value

**Comment by seanbudd** [comment]
Could you please update the title to be more descriptive of the change?

**Comment by nvdaes** [comment]
@leonardder and @seanbudd agree about this:
> Isn't this block executed on every focus change? I guess that's problematic.

I think that we can use the globalVars.oldFocusObject and create a variable with its treeInterceptor, if the old object exist. Then, if focusObject.treeInterceptor is not equal to the oldTreeInterceptor, then we can trigger the extension point. Otherwise, it should be triggered just when a tree interceptor exist and the the _set_passThrough function can take charge of this. I'll test this in a few hours after job.

**Comment by leonardder** [comment]
This sounds good to me. It might be good to have a look into reportPassThrough as well. May be that can be simplified based on this new extension point.

**Comment by nvdaes** [comment]
> It might be good to have a look into reportPassThrough as well. May be that can be simplified based on this new extension point.

Yes, though for now my tests are failing. Maybe in a new PR? Or, if you have suggestion, let me know please. I'll mark this as ready for review.

File: source/treeInterceptorHandler.py [hunk 3]
```diff
@@ -134,6 +144,8 @@ def _set_passThrough(self, state):
 		if self._passThrough == state:
 			return
 		self._passThrough = state
+		browseMode = not self._passThrough
```
Comment (seanbudd) [inline source/treeInterceptorHandler.py:147]: why does this use the underscored `_passThrough` but the changes in `api.py` uses `self.passThrough`

**Comment by nvdaes** [comment]
> why does this use the underscored _passThrough but the changes in api.py uses self.passThrough

My mistake. Fixed.

**Review by seanbudd — APPROVED** [review]
Thanks @nvdaes

**Pull request merged**